python3Packages.mypy-zope: init 0.3.2

superherointj opened the pull request:
MyPy here no longer works:

mypy/modulefinder.py:440: error: Argument 1 to "load" has incompatible type "TextIO"; expected "BinaryIO"
mypy/config_parser.py:172: error: Argument 1 to "load" has incompatible type "TextIO"; expected "BinaryIO"

SuperSandro hinted that:
They changed the type of a variable.
The patch needs to be adapted to that.

---

Closes #139319 Closes #130173 Closes #134312 Closes #139381

python3Packages:
* hydra-check: fix for mypy unstable-2021-09-17 
* types-typed-ast: init 1.4.4
* types-toml: init 0.10.0
* **mypy: 0.812 -> 0.910**
* mypy-zope: init 0.3.2
  - `mypy-zope` is a dependency of sygnal (issue #137092)

Upkeep:
* Add nicoo to the maintainers list
* types-{toml, typed-ast}: Add nicoo as maintainer

### Downstream packages fixed

* sublime-music: Skiped test "test_json_load_unload". Builds now.

* python{38,39}Packages.scancode-toolkit
  * Error
    ```
    FAILED tests/test_archive.py::TestRar::test_extract_rar_with_trailing_data - ...
    XFAIL tests/test_archive.py::TestUncompressGzip::test_uncompress_gzip_with_trailing_data
      Fails for now on Python 3
    ```
  * The fix was to:
    - Fix: `extractcode`
    - Upgrade: `commoncode`, `debian-inspector`, `license-expression`
    - Add: `parameter-expansion-patched`, `pygmars`
    - Minor refactoring (alphabetical order) on arguments. Don't mind!

* python{38,39}Packages.pylsp-mypy
  - Error 
    ```
    FAILED test/test_plugin.py::test_multiple_workspaces
    ```
  - Fix: Skip the test. :-)

* python3Packages.pyls-mypy:
  - Error: ``` error: python-language-server-0.36.2 not supported for interpreter python3.9 ```
  - Fix:
    - Skip 2 tests at `python-language-server`. 
    - `python-language-server` [does not support python39](https://github.com/palantir/python-language-server/issues/896#issuecomment-752790868). So `pyls-mypy` only builds using Python38 (or lower).
 
--

### Downstream packages still broken (Won't fix)

* python3Packages.algebraic-data-types
  - Error: ``` AttributeError: module 'mypy.types' has no attribute 'TypeVarDef' ```

nicoonoclaste committed:
Add nicoo to the maintainers list

superherointj committed:
python3Packages.mypy-zope: init 0.3.2

superherointj committed:
python3Packages.mypy: 0.812 -> unstable-2021-09-17

Co-authored-by: @jnetod @veehaitch @nbraud

superherointj committed:
python3Packages.types-toml: init 0.10.0

Co-authored-by: @jnetod @veehaitch @nbraud

superherointj committed:
python3Packages.types-typed-ast: init 1.4.4

Co-authored-by: @jnetod @veehaitch @nbraud

superherointj commented on pkgs/development/python-modules/mypy/default.nix:
```diff
@@ -1,28 +1,57 @@
   pname = "mypy";
-  version = "0.812";
+  version = "0.910";
   disabled = !isPy3k;
```
Sorry. Fixed.

SuperSandro2000 commented:
hydra-check fails now with the following:

```
hydracheck/cli.py:26: error:(B Library stubs not installed for (B"requests"(B (or incompatible with Python 3.9)(B
hydracheck/cli.py:26: note:(B Hint: (B"python3 -m pip install types-requests"(B(B
hydracheck/cli.py:26: note:(B (or run (B"mypy --install-types"(B to install all missing stub packages)(B
hydracheck/cli.py:26: note:(B See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports(B(B
```

jnetod commented:
It's a breaking change introduced in 0.900. We'll need to package some additional packages.

> Mypy now only ships with type stubs for stdlib modules (plus typing_extensions and mypy_extensions). If you use third-party libraries that don’t include stubs or inline type annotations, you can explicitly install stub packages, usually called types-<name>. 

https://mypy-lang.blogspot.com/2021/06/mypy-0900-released.html

SuperSandro2000 commented:
Can we just include those package in mypy by default? Then the upgrade should be safer.

nbraud commented:
> It's a breaking change introduced in 0.900. We'll need to package some additional packages.
> 
> > Mypy now only ships with type stubs for stdlib modules (plus typing_extensions and mypy_extensions). If you use third-party libraries that don’t include stubs or inline type annotations, you can explicitly install stub packages, usually called types-.
> 
> https://mypy-lang.blogspot.com/2021/06/mypy-0900-released.html

Yes, I was packaging the new dependencies and got blocked by what looks (possibly) like a `mypy` bug (it wasn't finding the type hint files even though they were installed, might only be searching the first folder in `$PYTHONPATH` or such) but @SuperSandro2000 closed the PR where I was doing that  >_>'

superherointj commented:
> Yes, I was packaging the new dependencies and got blocked by what looks (possibly) like a `mypy` bug (it wasn't finding the type hint files even though they were installed, might only be searching the first folder in `$PYTHONPATH` or such)

Is there any upstream report? `pyEnv` workaround worked for `hydra-check`.

> but @SuperSandro2000 closed the PR where I was doing that >_>'

Realize we are just trying to sort things out. If you have a better understanding of the situation, just convey it, and we will do our best to understand and integrate the necessary changes. If your PR had anything not included here just point out what it is and we'll fix it. Your contribution here is appreciated and welcomed.

nbraud commented:
> > Yes, I was packaging the new dependencies and got blocked by what looks (possibly) like a `mypy` bug (it wasn't finding the type hint files even though they were installed, might only be searching the first folder in `$PYTHONPATH` or such)
> 
> Is there any upstream report?

Not yet, was trying to make sure this was a `mypy` bug.

> `pyEnv` workaround worked for `hydra-check`.

Yeah, this looks like a valid workaround, at least in terms of getting `mypy` itself and `mypy-zope` to build.

> Realize we are just trying to sort things out. [...] If your PR had anything not included here just point out what it is and we'll fix it. Your contribution here is appreciated and welcomed.

Yes, I do appreciate that and I'm quite happy you picked things up.  I'm just concerned about the update breaking things for other packages that rely on `mypy` and do not use the `pyEnv` workaround; I will try to test that shortly.

I apologise if I came across as slighted or grumpy, it was before my morning coffee and I hadn't yet read the entire PR.  In any case, feel free to add me as co-maintainer on the `types-*` packages if you want the extra help (I had offered @veehaitch)

superherointj commented:
@nbraud I'll add you as co-maintainer shortly. Thanks for being understanding.

@jnetod found an upstream issue:
https://github.com/python/mypy/pull/11143

SuperSandro2000 commented:
@superherointj can you give those patches a try and see if we can remove the workaround again?

----

in the meantime I build home-assistant and nixpkgs-review successful. I think the potential breakage of this PR is very small and if the patch works correct we can merge this.

superherointj commented:
> @superherointj can you give those patches a try and see if we can remove the workaround again?
I'm on it.

> in the meantime I build home-assistant and nixpkgs-review successful. I think the potential breakage of this PR is very small and if the patch works correct we can merge this.
Agreed. But let me try the patches first. :)

nbraud commented:
Nice, thanks checking this out faster than I could @superherointj

SuperSandro2000 commented on pkgs/development/python-modules/mypy/default.nix:
```diff
@@ -1,28 +1,62 @@
+  propagatedBuildInputs = [
+    mypy-extensions
+    psutil
+    toml
+    tomli
```
You are sure we need both? Maybe a comment could clarify why this is needed.

superherointj replied:
It errored without it. Try removing it and you'll see. Unfortunately I'm not a Python programmer so I'm not sure of most things...

SuperSandro2000 commented on pkgs/development/python-modules/mypy/default.nix:
```diff
@@ -1,28 +1,62 @@
   pname = "mypy";
-  version = "0.812";
-  disabled = !isPy3k;
+  version = "2021-09-17";
```
```suggestion
  version = "unstable-2021-09-17";
```

superherointj replied:
Done

nbraud commented:
> Tried adding you but you are not listed as maintainer yet. Once this PR is merged, add yourself to the maintainer list and to the packages.

Ah, right, the commit adding me to the list was in my PR.  Sending you a patch for that.

nicoonoclaste committed:
types-{toml, typed-ast}: Add nicoo as maintainer

At superherointj's invitation:
	https://github.com/NixOS/nixpkgs/pull/140627#issuecomment-936961233

nbraud commented:
> > Ah, right, the commit adding me to the list was in my PR. Sending you a patch for that.
> 
> I think would be best if you added yourself.

Yeap, sent you [a PR] against this branch, so the relevant commits are all from me.

[a PR]: https://github.com/superherointj/nixpkgs/pull/3

superherointj committed:
hydra-check: fix for mypy unstable-2021-09-17

superherointj committed:
python3Packages.extractcode: skip test test_extract_rar_with_trailing_data

superherointj committed:
sublime-music: skip test test_json_load_unload

superherointj committed:
python3Packages.pylsp_mypy: disable test test_multiple_workspaces

superherointj commented:
I failed to fix python3Packages.{algebraic-data-types, pyls-mypy}
I'm not a Python programmer & I'm out of ideas to fix it.
*It's a good moment for someone else more competent to step in.*
I have no other changes in mind.
Feel free to review/merge this PR.

superherointj committed:
python38Packages.python-language-server: disable 2 tests

superherointj commented:
> Also please drop the hypothesis update again. It is a core part of the python eco system and causes a mass rebuild and probably breakages.

Sorry about the `hypothesis` update. You are very right here.

SuperSandro2000 commented:
> Should I mark it as broken package?

> python3Packages.python-language-server which does not support python39

Please mark python-language-server as disabled if python is at least 3.9. I think the following snipped should do that:

```nix
disabled = pythoAtLeast "3.9";
```

cole-h commented:
Is there a reason this is one big PR instead of multiple smaller PRs? Do all of these packages rely on each other?

SuperSandro2000 reviewed (changes requested):
Not sure if we should delay the pylint update together with asteroid.

SuperSandro2000 commented:
@superherointj I hope you saw that python-unstable is bumping a whole bunch of packages that you also bump here. If @mweinelt missed some please let me or him know that we can bump it while we are at it and help this effort here. Maybe we can also target python-unstable with this PR and fix there the remaining build errors?

@mweinelt what do you think?

superherointj commented:
I would like assistance to fix MyPy here. Errors as:

> mypy/modulefinder.py:440: error: Argument 1 to "load" has incompatible type "TextIO"; expected "BinaryIO"

> mypy/config_parser.py:172: error: Argument 1 to "load" has incompatible type "TextIO"; expected "BinaryIO"

superherointj commented:
> Hey @superherointj , just wanted to say thanks for your effort. I fell into a similar rabbit hole last time i tried to update "a small library" . Keep it up!

@makefu, it seems your congratulation was too early. I haven't solved this. Someone will need to pick this up and fix the patch. Sorry for that.
